modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/SessionNode.java

```diff
@@ -1131,14 +1131,20 @@ public ChildReferences getChildReferences( NodeCache cache ) {
     }
 
     private boolean allowsSNS( NodeCache cache ) {
-        Name primaryType = getPrimaryType(cache);
-        Set<Name> mixinTypes = getMixinTypes(cache);
         RepositoryEnvironment repositoryEnvironment = workspace(cache).repositoryEnvironment();
         if ( repositoryEnvironment == null) {
             return true;
         }
         NodeTypes nodeTypes = repositoryEnvironment.nodeTypes();
-        return nodeTypes == null || nodeTypes.allowsNameSiblings(primaryType, mixinTypes);
+        if (nodeTypes == null) {
+            return true;
+        }
+        Name primaryType = getPrimaryType(cache);
+        if (nodeTypes.isUnorderedCollection(primaryType)) {
+            return false;
+        }
+        Set<Name> mixinTypes = getMixinTypes(cache);
+        return nodeTypes.allowsNameSiblings(primaryType, mixinTypes);
```

Comment:
Another option might have been to change `NodeTypes.allowsNameSiblings` to also check whether unordered, since all unordered collections do not allow SNS. That might be better only if the logic above is repeated elsewhere.